Enhance MQTT cover platform

thecode opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
MQTT cover `value_template` is deprecated to use for extracting position, support will be removed in Home Assistant 2021.6.
Instead of using `value_template`, `position_template` should be used.

MQTT cover `tilt_invert_state` is deprecated, instead `tilt_min` and `tilt_max` should be used. Support for `tilt_invert_state` will be removed in Home Assistant 2021.6.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
### Summary
- Allow combining position and state in MQTT cover platform (based on https://github.com/home-assistant/core/pull/43949).
- Add template and fix optimistic in set tilt position.
- Add tests.

### Details
This change will allow combining `state_topic` and `position_topic` for covers which supports both state an position to allow setting cover position and showing the cover state which it is moving. In current implementation it is not possible to use both topics so if position is used it is not possible to show the cover state while it is moving.
In addition this PR fixes missing options for tilt to allow templating using `tilt_status_template` and fixes set tilt position to follow `tilt_optimistic` option correclty.

After discussion with community members we reached a conclusion that it would be clean to add an option for a cover to report a stopped state, Some covers only have 3 states [`opening`, `closing`, `stopped`] and not 4 states [`opening`, `closing`, `closed`, `opened`]
I have added an option to use the payload `stopped` in state topic. When this payload is received if the cover is in optimistic mode or position topic is not set the cover `open` or `closed` state will be set according to last direction the cover was moving.
If position topic is set the last position will be used to set the `open` or `closed` state.
This will allow clean implantations with such devices, for example Shelly 2.x and many "dumb" gate controllers which only have 3 states. This also aligns the state status topics with command status topics so for a `stop` command there is also a `stopped`
status.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [X] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
cover:
    # Combine state and position topics #
  - platform: mqtt
    name: "MQTT Cover"
    command_topic: "home-assistant/cover/set"
    state_topic: "home-assistant/cover/state"
    set_position_topic: "home-assistant/cover/postion/set"
    position_topic: "home-assistant/cover/postion"
    availability:
      - topic: "home-assistant/cover/availability"
    qos: 0
    retain: true
    payload_open: "OPEN"
    payload_close: "CLOSE"
    payload_stop: "STOP"
    state_open: "open"
    state_opening: "opening"
    state_closed: "closed"
    state_closing: "closing"
    payload_available: "online"
    payload_not_available: "offline"
    optimistic: false
    value_template: "{{ value.x }}"
    position_template: "{{ value.y }}"
       
    # Shelly 2.5 roller mode example #
  - platform: mqtt
    name: "Bedroom"
    command_topic: "shellies/shellyswitch25-ABCDEF/roller/0/command"
    state_topic: "shellies/shellyswitch25-ABCDEF/roller/0"
    set_position_topic: "shellies/shellyswitch25-ABCDEF/roller/0/command/pos"
    position_topic: "shellies/shellyswitch25-ABCDEF/roller/0/pos"
    state_closing: "close"
    state_opening: "open"
    state_stopped: "stop"
    qos: 0
    payload_open: "open"
    payload_close: "close"
    payload_stop: "stop"
    optimistic: false
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes https://github.com/home-assistant/core/issues/46014
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16466

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

thecode committed:
Enhance MQTT cover platform

Allow combining of position and state of MQTT cover
Add template and fix optimistic in set tilt position
Add tests

thecode committed:
Add abbreviations

emontnemery commented:
I think there should be tests verifying that:
- Without a state topic, state is set to closed/open when receiving position
- With a state topic, state is NOT set to closed/open when receiving position

Also, is there a need for some exception to set the state to closed/open when we get an update of the position to fully opened or fully closed?

thecode commented:
> I think there should be tests verifying that:
> 
> * Without a state topic, state is set to closed/open when receiving position

This is covered in few tests, for example: https://github.com/home-assistant/core/blob/08163a848c3a063f30260d4a24f86310bf197fd6/tests/components/mqtt/test_cover.py#L189

> * With a state topic, state is NOT set to closed/open when receiving position

Added test for this
> Also, is there a need for some exception to set the state to closed/open when we get an update of the position to fully opened or fully closed?

No, removed


After discussion with community members we reached a conclusion that it would be clean to add an option for a cover to report a `stopped` state, Some covers only have 3 states [`opening`, `closing`, `stopped`] and not 4 states [`opening`, `closing`, `closed`, `opened`]
I have added an option to use the payload `stopped` in state topic. When this payload is received if the cover is in optimistic mode or position topic is not set the cover `open` or `closed` state will be set according to last direction the cover was moving.
If position topic is set the last position will be used to set the `open` or `closed` state. 
This will allow clean implention with such devices, for example Shelly 2.x and many "dumb" gate controllers which only have 3 states. This also aligns the state status topics with command status topics so for a `stop` command there is also a `stopped` 

status.

thecode committed:
Add tests and stopped state

emontnemery commented:
Maybe we should change the behavior when calculating the position to send to the device with the following change in `async_set_cover_position`:
```diff
+        position = self.find_in_range_from_percent(position, COVER_PAYLOAD)
         if set_position_template is not None:
             position = set_position_template.async_render(parse_result=False, **kwargs)
-        else:
-            position = self.find_in_range_from_percent(position, COVER_PAYLOAD)
```


Background here: https://github.com/home-assistant/core/issues/46104#issuecomment-774492396

thecode committed:
Cleanup & fix range for templates

leroyloren commented:
I added a device for opening blinds to HA. Everything works fine, but the log is filled:

`WARNING (MainThread) [homeassistant.components.mqtt.cover] Payload is not integer within range:`

There are several hundred of these lines.

thecode commented:
> I added a device for opening blinds to HA. Everything works fine, but the log is filled:
> 
> `WARNING (MainThread) [homeassistant.components.mqtt.cover] Payload is not integer within range:`
> 
> There are several hundred of these lines.
Did you use the changes from  this PR?
Please show your cover MQTT configuration yaml.

leroyloren commented:
I only use Z2M integration, no additional configuration. I'll wait for the beta and then I'll get back to you. Thank you,

bieniu commented:
> I only use Z2M integration

So please show us payload of `config` topic for this cover from MQTT broker.

emontnemery commented:
@leroyloren Please don't spam PRs with unrelated issues, I'll mark your comments off topic to make the comments about the PR easier to follow. If you still have problems with zigbee2mqtt, please open a new issue for that.

emontnemery commented on homeassistant/components/mqtt/cover.py:
```diff
@@ -286,9 +329,16 @@ def state_message_received(msg):
+
+            template = self._config.get(CONF_GET_POSITION_TEMPLATE)
+
+            # To be deprecated:
```
```suggestion
            # To be removed in 2021.6:
```

emontnemery commented on tests/components/mqtt/test_cover.py:
```diff
@@ -1978,3 +2030,298 @@ async def test_entity_debug_info_message(hass, mqtt_mock):
+    ) in caplog.text
+
+
+async def test_no_deprecated_warnning_for_position_topic_using_position_template(
```
```suggestion
async def test_no_deprecated_warning_for_position_topic_using_position_template(
```

emontnemery commented on tests/components/mqtt/test_cover.py:
```diff
@@ -1978,3 +2030,298 @@ async def test_entity_debug_info_message(hass, mqtt_mock):
+async def test_no_deprecated_warnning_for_position_topic_using_position_template(
+    hass, caplog, mqtt_mock
+):
+    """Test no warnning when position_template is used for position_topic."""
```
```suggestion
    """Test no warning when position_template is used for position_topic."""
```

emontnemery committed:
Apply suggestions from code review
